bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/LoginManager.java

```diff
@@ -138,31 +150,67 @@ public void closeSession() throws FreeboxException {
         post(LOGOUT);
     }
 
-    private Status trackAuthorize(int trackId) throws FreeboxException {
-        return getSingle(AuthStatus.class, AUTHORIZE_ACTION, Integer.toString(trackId)).status();
+    private class GrantLoop implements Runnable {
+        private final ZonedDateTime timeLimit;
+        private final Authorization authorize;
+        private Status track;
+
+        public GrantLoop(ZonedDateTime timeLimit, Authorization authorize, Status track) {
+            this.timeLimit = timeLimit;
+            this.authorize = authorize;
+            this.track = track;
+        }
+
+        @Override
+        public void run() {
+            try {
+                track = ZonedDateTime.now().isAfter(timeLimit) ? Status.TIMEOUT
+                        : getSingle(AuthStatus.class, AUTHORIZE_ACTION, Integer.toString(authorize.trackId)).status();
+                if (track == Status.PENDING) {
+                    grantJob = Optional.of(scheduler.schedule(new GrantLoop(timeLimit, authorize, track), 2000,
+                            TimeUnit.MILLISECONDS));
+                    return;
+                }
+            } catch (FreeboxException e) {
+                logger.warn("Error during granting process : {}", e.getMessage());
+                track = Status.UNKNOWN;
+            }
+            grantJob = Optional.empty();
+        }
     }
 
     public String grant() throws FreeboxException {
+        ZonedDateTime timeLimit = ZonedDateTime.now().plusSeconds(GRANT_DELAY_SEC);
         Authorization authorize = post(new AuthorizeData(APP_ID, BUNDLE), AuthResponse.class, AUTHORIZE_ACTION);
         Status track = Status.PENDING;
-        ZonedDateTime timeLimit = ZonedDateTime.now().plusSeconds(GRANT_DELAY_SEC);
-        try {
-            while (Status.PENDING.equals(track) && ZonedDateTime.now().isBefore(timeLimit)) {
-                Thread.sleep(2000);
-                track = trackAuthorize(authorize.trackId());
-            }
-            switch (track) {
-                case GRANTED:
-                    return authorize.appToken();
-                case TIMEOUT, PENDING:
-                    throw new FreeboxException("Unable to grant session, delay expired");
-                case DENIED:
-                    throw new FreeboxException("Unable to grant session, access was denied");
-                default:
-                    throw new FreeboxException("Unable to grant session");
+
+        grantJob = Optional
+                .of(scheduler.schedule(new GrantLoop(timeLimit, authorize, track), 2000, TimeUnit.MILLISECONDS));
+
+        while (track == Status.PENDING) {
+            try {
+                grantJob.wait();
```

Comment:
Are you sure this is working? You're calling `wait()` on the Optional. However, `wait()` shouldn't be used directly, and if so, it must be called in a loop. https://wiki.sei.cmu.edu/confluence/display/java/THI03-J.+Always+invoke+wait%28%29+and+await%28%29+methods+inside+a+loop You could call `get()` on the `ScheduledFuture`.